### PR TITLE
Add ChatGLM model.

### PR DESCRIPTION
**Pull request by xunkai55**
# What does this PR do?

Add ChatGLM model support in HF Transformers repo.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@ArthurZucker

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! Thanks for opening a PR 🤗 I think the implementation of chatGLM should follow the comment I posted [here](https://github.com/huggingface/transformers/pull/27267#issuecomment-1802207501). See #27267 which also wanted to add support for ChatGM 😉

**Comment by nidhishgajjar** [comment]
> **Orb Code Review** (powered by GLM 5.1 on [Orb Cloud](https://orbcloud.dev))

Thanks for adding ChatGLM model support! This is a substantial contribution. I've reviewed the modeling, configuration, tokenizer, and test code. There are some significant issues that should be addressed.

---

### 🔴 Critical Security: `eval()` on model output in `process_response`

```python
def process_response(self, output, history):
    ...
    parameters = eval(content)
```

This executes arbitrary Python code from model-generated text. If the model produces malicious output, this is a **remote code execution vulnerability**. Replace with `json.loads()` or `ast.literal_eval()`, or at minimum document the security implications clearly.

---

### 🐛 Must-Fix: Global PyTorch settings modified at import time

```python
if sys.platform != "darwin":
    torch._C._jit_set_profiling_mode(False)
    torch._C._jit_set_profiling_executor(False)
    torch._C._jit_override_can_fuse_on_cpu(True)
    torch._C._jit_override_can_fuse_on_gpu(True)
```

These modify **global** PyTorch JIT settings when the module is imported, affecting every other model in the same process. This should either be removed, made opt-in, or scoped to only affect ChatGLM operations.

---

### ⚠️ Duplicated generation logic in `stream_generate`

The `stream_generate` method (~70 lines) reimplements most of the standard `generate()` logic (logits processing, stopping criteria, EOS handling, etc.). This creates significant maintenance burden and may miss bug fixes applied to the base class. Consider integrating with the standard `generate()` method using a custom `generate` callback or hook instead.

---

### ⚠️ `_init_weights` is effectively a no-op for normal usage

```python
def _init_weights(self, module: nn.Module):
    if not self.config.test_force_initializer_range:
        return
```

This means weight initialization only happens during testing. When creating a model from scratch (not from pretrained), weights won't be properly initialized. The standard pattern in transformers is to always initialize, with or without `test_force_initializer_range`.

---

### ⚠️ `InvalidScoreLogitsProcessor` uses magic numbers

```python
scores.zero_()
scores[..., 5] = 5e4
```

Token index `5` and score `5e4` are hardcoded without explanation. These should be documented or made configurable.

---

### ⚠️ Test design issues

- Tests reference `encoder_hidden_states` and `encoder_attention_mask` parameters, but ChatGLM is a decoder-only model that doesn't support cross-attention. These appear to be copied from an encoder-decoder test template and will fail.
- The test config passes `is_decoder=False` and `max_position_embeddings` which aren't in `ChatGLMConfig.__init__` parameters (they'd be silently absorbed by `**kwargs`).

---

### 📝 Minor Notes

- Both `import copy` and `from copy import deepcopy` are used — pick one for consistency
- `ChatGLMForSequenceClassification` pools using `hidden_states[-1]` which in the `s-b-h` layout is the last *sequence position*, not the last layer — worth a comment clarifying this is intentional
- The tokenizer properties (`unk_token`, `pad_token`, `eos_token`) are defined as `@property` but the superclass typically expects these to be settable — this may cause issues with serialization/deserialization

---

### Summary

The ChatGLM implementation is comprehensive with proper model architecture, tokenizer, configuration, and tests. The critical `eval()` call must be replaced with a safe alternative, and the global PyTorch JIT settings modification should be scoped properly. The custom generation loop adds maintenance risk that could be mitigated by leveraging the base class more.

### Assessment: request-changes

_This review was created by an AI assistant (OpenHands) on behalf of Orb Cloud._
